aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,645 @@
+      pools.begin(),
+      [indices_accessor, strides_ptr, dim, ndim] __device__(int64_t x) {
+        int64_t pool_index = 0;
+        for (int64_t j = 0; j < ndim; j++) {
+          if (j != dim) {
+            auto indices_row = indices_accessor[j];
+            auto stride = strides_ptr[j];
+            pool_index += stride * indices_row[x];
+          }
+        }
+        return pool_index;
+      });
+  int64_t* pools_ptr = pools.data().get();
+
+  thrust::device_vector<int64_t> sorted_indices(nnz);
+  thrust::sequence(sorted_indices.begin(), sorted_indices.end(), 0);
+
+  thrust::sort(
+      sorted_indices.begin(),
+      sorted_indices.end(),
+      [pools_ptr] __device__(int64_t x, int64_t y) {
+        return pools_ptr[x] < pools_ptr[y];
+      });
+  thrust::device_vector<int64_t> count_set(nnz);
+  thrust::device_vector<scalar_t*> mx_rows;
+  thrust::device_vector<scalar_t> log_softmax_max_buffer;
+  if (requireMxRows) {
+    thrust::device_vector<int64_t> mx_rows_indices(nnz);
+    auto values_accessor =
+        values.packed_accessor<scalar_t, 2>(); // {nnz, nvalues}
+
+    auto val_zip_ter = thrust::make_zip_iterator(thrust::make_tuple(
+        sorted_indices.begin(), thrust::make_constant_iterator(int64_t(1))));
+    auto out_val_zip_ter = thrust::make_zip_iterator(
+        thrust::make_tuple(mx_rows_indices.begin(), count_set.begin()));
+
+    auto new_end = thrust::reduce_by_key(
+        sorted_indices.begin(),
+        sorted_indices.end(),
+        val_zip_ter,
+        thrust::make_discard_iterator(),
+        out_val_zip_ter,
+        [pools_ptr] __device__(int64_t x, int64_t y) {
+          return pools_ptr[x] == pools_ptr[y];
+        },
+        [nvalues, values_accessor] __device__(
+            thrust::tuple<int64_t, int64_t> x,
+            thrust::tuple<int64_t, int64_t> y) {
+          auto index_x = thrust::get<0>(x);
+          auto count_x = thrust::get<1>(x);
+
+          auto index_y = thrust::get<0>(y);
+          auto count_y = thrust::get<1>(y);
+          const scalar_t* a = values_accessor[index_x].data();
+          const scalar_t* b = values_accessor[index_y].data();
+
+          int64_t j = 0;
+          while (j < nvalues && a[j] > b[j]) {
+            j++;
+          }
+          if (j == nvalues)
+            return thrust::make_tuple(index_x, count_x + count_y);
+          return thrust::make_tuple(index_y, count_x + count_y);
+        });
+    auto iter_tuple = new_end.second.get_iterator_tuple();
+    auto new_sz =
+        thrust::distance(mx_rows_indices.begin(), thrust::get<0>(iter_tuple));
+    mx_rows_indices.resize(new_sz);
+    count_set.resize(new_sz);
+    mx_rows.resize(new_sz);
+    if (LogSoftMax) {
+      log_softmax_max_buffer.resize(new_sz * nvalues);
+      auto log_softmax_max_buffer_ptr = log_softmax_max_buffer.data().get();
+      for (int64_t index = 0; index < new_sz; index++) {
+        auto current = mx_rows_indices[index];
+        cudaMemcpy(
```

Comment:
please don't use synchronizing cuda APIs. The best way would be to use pytorch's APIs to copy tensors, if this is undesirable for some reason, use cudaMemcpyAsync and specify stream it has to run on.

Reply:
all right!